### PR TITLE
Implemented API for storing and getting scan details

### PR DESCRIPTION
**Pull request by alex-ko-dev**
### Description

The APIs are required for the interaction between the new ast-flow-listener and ast-flow-publisher services.

### References

Part of work item [516](https://dev.azure.com/CxFlow/CxFlow/_workitems/edit/516).

### Testing

See scan-details-api.feature

Request examples can also be found in datastore.http ("scan details" section).

### CODE REVIEW

**Comment by cxflowtestuser** [comment]
Scan submitted to Checkmarx

**Comment by cxflowtestuser** [comment]
Logo<br />**Checkmarx SAST - Scan Summary & Details**
### Cx-SAST Summary
Total of 105 vulnerabilities<br />High&nbsp;**2 High**<br />Medium&nbsp;**8 Medium**<br />Low&nbsp;**92 Low**<br />Info&nbsp;**3 Info**<br />
### Violation Summary
High&nbsp;**2** **High**<br />Medium&nbsp;**5** **Medium**<br />
[View more details on Checkmarx UI](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1052167&projectid=1182)
### Cx-SAST Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---
[53](https://github.com/checkmarx-ltd/Integrations-Datastore/blob/alexey-516-save-scan-job/src/main/java/com/checkmarx/integrations/datastore/controllers/RepoController.java#L53) |Medium|Unsafe_Object_Binding|src/main/java/com/checkmarx/integrations/datastore/controllers/RepoController.java|[Checkmarx](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1052167&projectid=1182&pathid=12)
[47](https://github.com/checkmarx-ltd/Integrations-Datastore/blob/alexey-516-save-scan-job/src/main/java/com/checkmarx/integrations/datastore/controllers/ScmController.java#L47) |Medium|Unsafe_Object_Binding|src/main/java/com/checkmarx/integrations/datastore/controllers/ScmController.java|[Checkmarx](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1052167&projectid=1182&pathid=15)
[40](https://github.com/checkmarx-ltd/Integrations-Datastore/blob/alexey-516-save-scan-job/src/main/java/com/checkmarx/integrations/datastore/controllers/TokenController.java#L40) [52](https://github.com/checkmarx-ltd/Integrations-Datastore/blob/alexey-516-save-scan-job/src/main/java/com/checkmarx/integrations/datastore/controllers/TokenController.java#L52) |Medium|Unsafe_Object_Binding|src/main/java/com/checkmarx/integrations/datastore/controllers/TokenController.java|[Checkmarx](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1052167&projectid=1182&pathid=16)
[30](https://github.com/checkmarx-ltd/Integrations-Datastore/blob/alexey-516-save-scan-job/src/main/java/com/checkmarx/integrations/datastore/controllers/OrgController.java#L30) |Medium|Unsafe_Object_Binding|src/main/java/com/checkmarx/integrations/datastore/controllers/OrgController.java|[Checkmarx](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1052167&projectid=1182&pathid=10)
[27](https://github.com/checkmarx-ltd/Integrations-Datastore/blob/alexey-516-save-scan-job/src/main/java/com/checkmarx/integrations/datastore/services/RepoService.java#L27) |High|Stored_XSS|src/main/java/com/checkmarx/integrations/datastore/services/RepoService.java|[Checkmarx](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1052167&projectid=1182&pathid=4)
[27](https://github.com/checkmarx-ltd/Integrations-Datastore/blob/alexey-516-save-scan-job/src/main/java/com/checkmarx/integrations/datastore/controllers/RepoController.java#L27) |High|Reflected_XSS_All_Clients|src/main/java/com/checkmarx/integrations/datastore/controllers/RepoController.java|[Checkmarx](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1052167&projectid=1182&pathid=8)
***
Logo
**Checkmarx SCA - Scan Summary & Details**
### Cx-SCA Summary
**Total Packages Identified**: **126**
**Scan Risk Score**: **9.80**

High&nbsp;**4** **High** **severity vulnerabilities**
Medium&nbsp;**0** **Medium** **severity vulnerabilities**
Low&nbsp;**0** **Low** **severity vulnerabilities**
[View more details on Checkmarx UI](https://eu.sca.checkmarx.net/#/projects/b3924419-f855-4b95-aad9-7afd87841361/reports/c2178edd-f0c5-4067-bb97-8ef83d09aee7)
### Cx-SCA vulnerability result overview
|Vulnerability ID|Package|Severity|CVSS score|Publish date|Current version|Recommended version|Link in CxSCA|Reference – NVD link|
---|---|---|---|---|---|---|---|---
`CVE-2016-1000027`|org.springframework:spring-web|HIGH|9.8|2020-01-02T23:15:00|5.2.9.RELEASE|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/b3924419-f855-4b95-aad9-7afd87841361/reports/c2178edd-f0c5-4067-bb97-8ef83d09aee7/vulnerabilities/CVE-2016-1000027%3AYarn-org.springframework%3Aspring-web-5.2.9.RELEASE/vulnerabilityDetails)|[CVE-2016-1000027](https://nvd.nist.gov/vuln/detail/CVE-2016-1000027)
`CVE-2016-1000027`|org.springframework:spring-webmvc|HIGH|9.8|2020-01-02T23:15:00|5.2.9.RELEASE|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/b3924419-f855-4b95-aad9-7afd87841361/reports/c2178edd-f0c5-4067-bb97-8ef83d09aee7/vulnerabilities/CVE-2016-1000027%3AYarn-org.springframework%3Aspring-webmvc-5.2.9.RELEASE/vulnerabilityDetails)|[CVE-2016-1000027](https://nvd.nist.gov/vuln/detail/CVE-2016-1000027)
`CVE-2021-20190`|com.fasterxml.jackson.core:jackson-databind|HIGH|8.1|2021-01-19T17:15:00|2.11.2|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/b3924419-f855-4b95-aad9-7afd87841361/reports/c2178edd-f0c5-4067-bb97-8ef83d09aee7/vulnerabilities/CVE-2021-20190%3AYarn-com.fasterxml.jackson.core%3Ajackson-databind-2.11.2/vulnerabilityDetails)|[CVE-2021-20190](https://nvd.nist.gov/vuln/detail/CVE-2021-20190)
`CVE-2020-25638`|org.hibernate:hibernate-core|HIGH|7.4|2020-09-22T16:32:00|5.4.21.Final|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/b3924419-f855-4b95-aad9-7afd87841361/reports/c2178edd-f0c5-4067-bb97-8ef83d09aee7/vulnerabilities/CVE-2020-25638%3AYarn-org.hibernate%3Ahibernate-core-5.4.21.Final/vulnerabilityDetails)|[CVE-2020-25638](https://nvd.nist.gov/vuln/detail/CVE-2020-25638)

File: src/main/java/com/checkmarx/integrations/datastore/controllers/exceptions/GlobalControllerAdvisor.java [hunk 1]
```diff
@@ -17,22 +17,31 @@
 public class GlobalControllerAdvisor extends ResponseEntityExceptionHandler {
     @ExceptionHandler(EntityNotFoundException.class)
     public ResponseEntity<GeneralExceptionDO> handleRepoNotFoundException(EntityNotFoundException e) {
+        // There is no need to log a huge stack trace.
+        log.debug("Intercepted {}", e.toString());
+
```
[the remark this answers is not in the document]
Comment (alex-ko-dev) [reply src/main/java/com/checkmarx/integrations/datastore/controllers/exceptions/GlobalControllerAdvisor.java]: Fixed.

File: src/main/java/com/checkmarx/integrations/datastore/controllers/exceptions/GlobalControllerAdvisor.java [hunk 2]
```diff
@@ -17,22 +17,31 @@
...
         GeneralExceptionDO externalException = toExternalException(e);
         return new ResponseEntity<>(externalException, HttpStatus.NOT_FOUND);
     }
 
+    @ExceptionHandler(DuplicateKeyException.class)
+    public ResponseEntity<GeneralExceptionDO> handleDuplicateKeyException(DuplicateKeyException e) {
+        log.debug("Intercepted {}", e.toString());
+        return new ResponseEntity<>(toExternalException(e), HttpStatus.BAD_REQUEST);
```
[the remark this answers is not in the document]
Comment (alex-ko-dev) [reply src/main/java/com/checkmarx/integrations/datastore/controllers/exceptions/GlobalControllerAdvisor.java]: Fixed.

**Review by AsafCX — APPROVED** [review]

**Pull request merged**